subsys/bluetooth/services/ots/ots_dir_list.c

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/types.h>
+#include <stddef.h>
+#include <string.h>
+#include <errno.h>
+#include <sys/dlist.h>
+#include <sys/byteorder.h>
+
+#include <bluetooth/services/ots.h>
+#include "ots_internal.h"
+#include "ots_obj_manager_internal.h"
+
+#include <logging/log.h>
+
+LOG_MODULE_DECLARE(bt_ots, CONFIG_BT_OTS_LOG_LEVEL);
+
+static struct bt_ots_dir_list dir_lists[CONFIG_BT_OTS_MAX_INST_CNT];
+
+static void dir_list_object_encode(struct bt_gatt_ots_object *obj,
+				     struct net_buf_simple *net_buf)
+{
+	uint8_t flags = 0;
+	uint8_t *start;
+	uint16_t start_len = net_buf->len;
+
+	BT_OTS_DIR_LIST_SET_FLAG_PROPERTIES(flags);
+	BT_OTS_DIR_LIST_SET_FLAG_CUR_SIZE(flags);
+	if (obj->metadata.type.uuid.type == BT_UUID_TYPE_128) {
+		BT_OTS_DIR_LIST_SET_FLAG_TYPE_128(flags);
+	}
+
+	/* skip 16bits at the beginning of the record for the record's length */
+	start = net_buf_simple_add(net_buf, sizeof(uint16_t));
+
+	/* ID */
+	net_buf_simple_add_le48(net_buf, obj->id);
+
+	/* Name length */
+	net_buf_simple_add_u8(net_buf, strlen(obj->metadata.name));
+
+	/* Name */
+	net_buf_simple_add_mem(net_buf, obj->metadata.name,
+			       strlen(obj->metadata.name));
+
+	/* Flags */
+	net_buf_simple_add_u8(net_buf, flags);
+
+	/* encode Object type */
+	if (obj->metadata.type.uuid.type == BT_UUID_TYPE_16) {
+		net_buf_simple_add_le16(net_buf, obj->metadata.type.uuid_16.val);
+	} else {
+		net_buf_simple_add_mem(net_buf, obj->metadata.type.uuid_128.val,
+				       BT_UUID_SIZE_128);
+	}
+
+	/* encode Object Current size */
+	net_buf_simple_add_le32(net_buf, obj->metadata.size.cur);
+
+	/* Object properties */
+	net_buf_simple_add_le32(net_buf, obj->metadata.props);
+
+	/* Update the record length at the beginning */
+	sys_put_le16(net_buf_simple_tail(net_buf) - start, start);
```

Comment:
Nice, but now you're not using `start_len` anymore, i.e. you should probably remove it.

Btw, just double-checking: the length field is supposed to include the length of the length field itself and not only the data that follows it (that latter is more common in other Bluetooth protocols I'm familiar with)

Reply:
Right :D 

Just double checked the spec: 
> Each Object Record shall begin with a “Length of Object Record” field that specifies the total size of the current Object Record in octets. The total value reported includes the 2 octets used by the “Length of Object Record” field itself.

So that should be right :)